treewide: reconnect receiver and repeat last message

mafredri opened the pull request:
This PR adds support for reconnecting the WebSocket on error and implements a new feature to repeat the last message for `/receive/pop`.

Previously the API was left running but no new messages received if e.g. signal-cli-rest-api was restarted.

The repeat the last message feature is useful in Home Assistant to avoid clearing the sensor status on next scan.

Thanks for creating the project, came in handy. And I pretty much skipped writing tests for the new functionality so let me know if you want me to add those. Cheers.

kalbasit commented:
@mafredri thank you for your contribution, much appreciated! Do you mind splitting out the `repeat last message` feature into its own PR? Tests are much appreciated. Thank you!

mafredri commented:
@kalbasit I made the changes you requested but I see you've made a lot of changes since then that don't merge nicely.

In case you want it, here's the test I added for the reconnect feature:

```go
type mockClient struct {
	connectCalled int
	connectErr    chan error
	recvMsg       chan receiver.Message
	recvErr       chan error
	msgs          []receiver.Message
}

func (mc *mockClient) Connect() error {
	mc.connectCalled++
	if mc.connectErr != nil {
		return <-mc.connectErr
	}
	return nil
}

func (mc *mockClient) ReceiveLoop() error {
	for {
		select {
		case msg := <-mc.recvMsg:
			mc.msgs = append(mc.msgs, msg)
		case err := <-mc.recvErr:
			return err
		}
	}
}

func (mc *mockClient) Connect() error {
	return nil
}

func (mc *mockClient) ReceiveLoop() error {
	<-(make(chan struct{})) // block forever.

	return nil
}

func (mc *mockClient) Pop() *receiver.Message {
	if len(mc.msgs) == 0 {
		return nil
	}

	msg := mc.msgs[0]
	mc.msgs = mc.msgs[1:]

	return &msg
}

func (mc *mockClient) Flush() []receiver.Message {
	msgs := mc.msgs
	mc.msgs = []receiver.Message{}

	return msgs
}

func TestServerReconnect(t *testing.T) {
	mc := &mockClient{
		connectErr: make(chan error),
		recvMsg:    make(chan receiver.Message),
		recvErr:    make(chan error),
	}
	_ = New(mc)

	if want, got := 0, mc.connectCalled; want != got {
		t.Errorf("want %d, got %d", want, got)
	}

	mc.recvMsg <- receiver.Message{Account: "0"}

	mc.recvErr <- nil
	mc.connectErr <- nil

	if want, got := 1, len(mc.msgs); want != got {
		t.Errorf("len(mc.msgs) want %d, got %d", want, got)
	}

	mc.recvMsg <- receiver.Message{Account: "0"}

	if want, got := 1, mc.connectCalled; want != got {
		t.Errorf("mc.connectCalled want %d, got %d", want, got)
	}

	mc.recvErr <- nil

	if want, got := 2, len(mc.msgs); want != got {
		t.Errorf("len(mc.msgs) want %d, got %d", want, got)
	}
}
```

And here's what I added for repeat:

```go
func TestConnectAndReconnect(t *testing.T) {
	newMessage := func(i int) Message {
		message := strconv.Itoa(i)
		return Message{
			Envelope: Envelope{
				DataMessage: &DataMessage{
					Message: &message,
				},
			},
		}
	}

	ch := make(chan chan Message, 1)
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		upgrader := websocket.Upgrader{}
		conn, err := upgrader.Upgrade(w, r, nil)
		if err != nil {
			t.Errorf("upgrade websocket: %v", err)
			return
		}
		defer conn.Close()

		messages := <-ch
		for msg := range messages {
			if err := conn.WriteJSON(msg); err != nil {
				t.Errorf("write message: %v", err)
				return
			}
		}
	}))
	defer server.Close()

	uri, err := url.Parse("ws" + strings.TrimPrefix(server.URL, "http"))
	if err != nil {
		t.Fatalf("parse server URL: %v", err)
	}

	client, err := New(uri)
	if err != nil {
		t.Fatalf("New: %v", err)
	}

	for j := 1; j <= 2; j++ {
		messages := make(chan Message, 3)
		for i := 1; i <= cap(messages); i++ {
			messages <- newMessage(j * i)
		}
		close(messages)
		ch <- messages

		err = client.ReceiveLoop()
		if err == nil {
			t.Fatalf("ReceiveLoop: want an error, got nil")
		}

		for i := 1; i <= 3; i++ {
			msg := client.Pop()
			if msg == nil {
				t.Errorf("Pop: want a message, got nil")
			} else if want, got := strconv.Itoa(j*i), *msg.Envelope.DataMessage.Message; want != got {
				t.Errorf("msg.Account: want %q got %q", want, got)
			}
		}

		err = client.Connect()
		if err != nil {
			t.Fatalf("Connect: %v", err)
		}
	}
}
```

kalbasit commented:
@mafredri thank you! I opened #40 and will add these tests, appreciate it!
